src/app/archive/archive_lib/processor.ml

```diff
@@ -2125,8 +2151,326 @@ module Block_and_signed_command = struct
                FROM blocks_user_commands
                WHERE block_id = $1
                AND user_command_id = $2
+               AND sequence_no = $3
            |sql})
-      (block_id, user_command_id)
+      (block_id, user_command_id, sequence_no)
+end
+
+module Zkapp_party_failures = struct
+  type t = { index : int; failures : string array } [@@deriving fields, hlist]
+
+  let typ =
+    Mina_caqti.Type_spec.custom_type ~to_hlist ~of_hlist
+      Caqti_type.[ int; Mina_caqti.array_string_typ ]
+
+  let table_name = "zkapp_party_failures"
+
+  let add_if_doesn't_exist (module Conn : CONNECTION) index failures =
+    let failures =
+      List.map failures ~f:Transaction_status.Failure.to_string |> Array.of_list
+    in
+    Mina_caqti.select_insert_into_cols ~select:("id", Caqti_type.int)
+      ~table_name
+      ~cols:([ "index"; "failures" ], typ)
+      (module Conn)
+      { index; failures }
+
+  let load (module Conn : CONNECTION) id =
+    Conn.find
+      (Caqti_request.find Caqti_type.int typ
+         (Mina_caqti.select_cols_from_id ~table_name
+            ~cols:[ "index"; "failures" ]))
+      id
+end
+
+module Block_and_zkapp_command = struct
+  type t =
+    { block_id : int
+    ; zkapp_command_id : int
+    ; sequence_no : int
+    ; status : string
+    ; failure_reasons_ids : int array option
+    }
+  [@@deriving hlist]
+
+  let table_name = "blocks_zkapp_commands"
+
+  let typ =
+    Mina_caqti.Type_spec.custom_type ~to_hlist ~of_hlist
+      Caqti_type.[ int; int; int; string; option Mina_caqti.array_int_typ ]
+
+  let add_if_doesn't_exist (module Conn : CONNECTION) ~block_id
+      ~zkapp_command_id ~sequence_no ~status
+      ~(failure_reasons : Transaction_status.Failure.Collection.t option) =
```

Comment:
It's better to use `Transaction_status.Failure.Collection.display`. The `display` type already has the index into the parties and it doesn't contains empty list. Everything else would be similar.
```
Mina_caqti.deferred_result_list_map reasons ~f:(fun (ndx, failure_reasons) ->
...
```